### PR TITLE
Updated to wgpu 0.16.0, wgpu-hal 0.16.0 and naga 0.12.0

### PR DESCRIPTION
**Pull request by airingursb**
# Objective

- Updated to wgpu 0.16.0 and wgpu-hal 0.16.0

---

## Changelog

1. Upgrade wgpu to 0.16.0 and  wgpu-hal to 0.16.0
2. Fix the error in native when using a filterable `TextureSampleType::Float` on a multisample `BindingType::Texture`. ([https://github.com/gfx-rs/wgpu/pull/3686](https://github.com/gfx-rs/wgpu/pull/3686))

### CODE REVIEW

**Comment by mockersf** [comment]
Could you also update naga to 0.12?

You'll have to change a few other things for examples to work, don't hesitate to run them all on your computer (`bash ./tools/example_showcase.sh`)

**Comment by airingursb** [comment]
> Could you also update naga to 0.12?
> 
> You may have to change a few other things for examples to work, don't hesitate to run them all on your computer (`bash ./tools/example_showcase.sh`). Example `shader_prepass` is failing for me without additional changes.

Okay, I will update naga and  then run the examples to verify.

File: crates/bevy_render/src/texture/fallback_image.rs [hunk 1]
```diff
@@ -54,15 +52,15 @@ fn fallback_image_new(
     image.texture_descriptor.usage |= TextureUsages::RENDER_ATTACHMENT;
 
     // We can't create textures with data when it's a depth texture or when using multiple samples
-    let texture = if format.describe().sample_type == TextureSampleType::Depth || samples > 1 {
+    let texture = if format.sample_type(None) == Some(TextureSampleType::Depth) || samples > 1 {
```
Comment (RobWalt) [inline crates/bevy_render/src/texture/fallback_image.rs:55]: Why is it ok to pass `None` here?

File: crates/bevy_render/src/texture/image.rs [hunk 2]
```diff
@@ -483,9 +483,9 @@ pub trait TextureFormatPixelInfo {
 
 impl TextureFormatPixelInfo for TextureFormat {
     fn pixel_size(&self) -> usize {
-        let info = self.describe();
-        match info.block_dimensions {
-            (1, 1) => info.block_size.into(),
+        let info = self;
+        match info.block_dimensions() {
+            (1, 1) => (info.block_size(None).unwrap_or_default() as usize).into(),
```
Comment (RobWalt) [inline crates/bevy_render/src/texture/image.rs:488]: Same question as above here. Additionally why can we use `unwrap_or_default` here?

File: crates/bevy_render/src/texture/ktx2.rs [hunk 3]
```diff
@@ -233,12 +233,12 @@ pub fn ktx2_buffer_to_image(
...
-    let texture_format_info = texture_format.describe();
+    let texture_format_info = texture_format;
     let (block_width_pixels, block_height_pixels) = (
-        texture_format_info.block_dimensions.0 as usize,
-        texture_format_info.block_dimensions.1 as usize,
+        texture_format_info.block_dimensions().0 as usize,
+        texture_format_info.block_dimensions().1 as usize,
     );
-    let block_bytes = texture_format_info.block_size as usize;
+    let block_bytes = texture_format_info.block_size(None).unwrap_or_default() as usize;
```
Comment (RobWalt) [inline crates/bevy_render/src/texture/ktx2.rs:241]: And same question here just for documentation purposes

**Review by RobWalt — COMMENTED** [review]
Hey cool! I was working on the same changes. I don't have any knowledge at all in the specifics of that domain and I was struggling at a few places. Can you maybe clarify a few changes?

Feel free to ignore my comments if they get in the way.

**Comment by Azorlogh** [comment]
Just a heads up for the changelog, this update also fixes this [issue 5732](https://github.com/bevyengine/bevy/issues/5732) :slightly_smiling_face:

File: crates/bevy_render/src/texture/fallback_image.rs (hunk 1 above)
Comment (airingursb) [reply crates/bevy_render/src/texture/fallback_image.rs:55]: Thank you for your sincere question. I don't know much about wgpu. These changes refer to [the API migration advice provided by changelog in wgpu 0.16](https://github.com/gfx-rs/wgpu/blob/v0.16/CHANGELOG.md#v0160-2023-04-19). 
If there is any inaccuracy here, please also ask Reviewer for advice.

**Review by devil-ira — APPROVED** [review]
LGTM. Changes match what's in the wgpu changelog.

**Comment by mockersf** [comment]
There are some code to fix behind features, and some formatting to update.

**Comment by RobWalt** [comment]
FYI, just a small warning:

We found some crashes with the update to wgpu 0.16. It only happens when compiling to wasm, running the app on MacOS and changing the perspective of the camera. It crashes not only the app, but the browser. We tested this on several browsers and it happened everywhere everytime. The App was running normally when compiled for native use. It's most likely some issue with MacOS + the new version of wgpu and isn't caused by bevy directly.

 We are still investigating it and are working on a minimal repro example to test if it might be related to that specific machine or if it's really related to MacOS in general.

**Comment by superdump** [comment]
Please run with `--features zstd,ktx2,dds` to catch a couple more errors. And then it needs to be run through `cargo fmt`. Other than that, if it's working then it looks good to me.

**Comment by airingursb** [comment]
> Please run with `--features zstd,ktx2,dds` to catch a couple more errors. And then it needs to be run through `cargo fmt`. Other than that, if it's working then it looks good to me.

@superdump I have revised your question over these commits. Please see if there are any other items that need to be modified.

- [6c59137](https://github.com/bevyengine/bevy/pull/8446/commits/6c59137ffcb5739a98b1975108492b849c663493) 
- [fad0a77](https://github.com/bevyengine/bevy/pull/8446/commits/fad0a775907e6eb30920b4df9641897ab4bf27f0)

File: crates/bevy_render/src/texture/fallback_image.rs (hunk 1 above)
Comment (mockersf) [inline crates/bevy_render/src/texture/fallback_image.rs:55]:
```suggestion
    let texture = if format.is_depth_stencil_format() || samples > 1 {
```

**Comment by mockersf** [comment]
Could you also compile with feature `basis-universal`? CI is failing on another `decribe()`

File: crates/bevy_render/src/texture/fallback_image.rs (hunk 1 above)
Comment (airingursb) [reply crates/bevy_render/src/texture/fallback_image.rs]: Updated

File: crates/bevy_render/src/texture/ktx2.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (airingursb) [reply crates/bevy_render/src/texture/ktx2.rs:242]: Updated

**Comment by airingursb** [comment]
> Could you also compile with feature `basis-universal`? CI is failing on another `decribe()`

@mockersf  Thanks to suggestion, the related CI problem has been fixed. I need your help to take a look at the version of the `Cargo.lock` file. The `check-bans` CI failed.

**Comment by mockersf** [comment]
I've looked at the `check-bans` failures, they can be ignored 👍

File: crates/bevy_render/src/texture/image.rs [hunk 4]
```diff
@@ -483,9 +483,9 @@ pub trait TextureFormatPixelInfo {
 
 impl TextureFormatPixelInfo for TextureFormat {
     fn pixel_size(&self) -> usize {
-        let info = self.describe();
-        match info.block_dimensions {
-            (1, 1) => info.block_size.into(),
+        let info = self;
+        match info.block_dimensions() {
+            (1, 1) => info.block_size(None).unwrap_or_default() as usize,
```
[the remark this answers is not in the document]
Comment (airingursb) [reply crates/bevy_render/src/texture/image.rs]: OK, updated at [639a09a](https://github.com/bevyengine/bevy/pull/8446/commits/639a09a802d90ee5d935410461d72e8bacedd081)

**Review by mockersf — APPROVED** [review]

File: crates/bevy_render/src/texture/image.rs [hunk 5]
```diff
@@ -387,15 +387,15 @@ impl Image {
 
     /// Whether the texture format is compressed or uncompressed
     pub fn is_compressed(&self) -> bool {
-        let format_description = self.texture_descriptor.format.describe();
+        let format_description = self.texture_descriptor.format;
         format_description
-            .required_features
-            .contains(wgpu::Features::TEXTURE_COMPRESSION_ASTC_LDR)
+            .required_features()
```
Comment (superdump) [inline crates/bevy_render/src/texture/image.rs:392]: Is `required_features()` cheap? I would guess so but we should check.

Comment (superdump) [reply crates/bevy_render/src/texture/image.rs:392]: Yes, cheap enough. It's a match over texture formats.

File: crates/bevy_render/src/texture/ktx2.rs [hunk 6]
```diff
@@ -154,12 +154,13 @@ pub fn ktx2_buffer_to_image(
                 TranscodeFormat::Uastc(data_format) => {
                     let (transcode_block_format, texture_format) =
                         get_transcoded_formats(supported_compressed_formats, data_format, is_srgb);
-                    let texture_format_info = texture_format.describe();
+                    let texture_format_info = texture_format;
                     let (block_width_pixels, block_height_pixels) = (
-                        texture_format_info.block_dimensions.0 as u32,
-                        texture_format_info.block_dimensions.1 as u32,
+                        texture_format_info.block_dimensions().0,
```
[the remark this answers is not in the document]
Comment (superdump) [reply crates/bevy_render/src/texture/ktx2.rs:159]: Yes, cheap enough. It's a match over texture formats.

**Pull request merged**